internal/handlers/pg/msg_count.go

```diff
@@ -85,37 +88,46 @@ func (h *Handler) MsgCount(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, e
 
 	resDocs := make([]*types.Document, 0, 16)
 	err = h.PgPool.InTransaction(ctx, func(tx pgx.Tx) error {
-		fetchedChan, err := h.PgPool.QueryDocuments(ctx, tx, &sp)
-		if err != nil {
-			return err
+		it, ierr := h.PgPool.GetDocuments(ctx, tx, &sp)
+		if ierr != nil {
+			return ierr
 		}
-		defer func() {
-			// Drain the channel to prevent leaking goroutines.
-			// TODO Offer a better design instead of channels: https://github.com/FerretDB/FerretDB/issues/898.
-			for range fetchedChan {
-			}
-		}()
 
-		for fetchedItem := range fetchedChan {
-			if fetchedItem.Err != nil {
-				return fetchedItem.Err
-			}
+		if it == nil {
+			// no documents found
+			return nil
+		}
 
-			for _, doc := range fetchedItem.Docs {
-				matches, err := common.FilterDocument(doc, filter)
-				if err != nil {
-					return err
-				}
+		for {
+			_, doc, ierr := it.Next()
+
+			switch {
+			case ierr == nil:
+				// do nothing
+			case errors.Is(ierr, iterator.ErrIteratorDone):
+				// no more documents
+				return nil
+			case errors.Is(ierr, context.Canceled), errors.Is(ierr, context.DeadlineExceeded):
```

Comment:
I think checking `ctx.Err()` above would be easier

Reply:
@AlekSi hmmm... where above? I probably don't understand something. The idea is that we pass `ctx` to new iterator, and at the beginning of each `Next` we check if the context is not closed. It has some point, but it's not super critical, at least with the current implementation. So, should I remove it or something?

Reply:
We should replace `case errors.Is(ierr, context.Canceled), errors.Is(ierr, context.DeadlineExceeded)` with `if ctx.Err() != nil` before switch. For example:

https://github.com/FerretDB/FerretDB/blob/8bc6b5dd2264d4c45ba8761ad98b1faa325f7bee/internal/clientconn/listener.go#L144-L156

We don't even inspect `err` if context was canceled

Reply:
fixed